util/log/file.go

```diff
@@ -221,3 +301,103 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
 }
+
+// int64sortable is required so we can sort on an int64 slice.
+type int64sortable []int64
+
+func (a int64sortable) Len() int           { return len(a) }
+func (a int64sortable) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a int64sortable) Less(i, j int) bool { return a[i] < a[j] }
+
+// selectFiles selects all log files that have an timestamp before the endTime and
+// the correct severity. It then sorts them in decreasing order, with the most
+// recent as the first one.
+func selectFiles(logFiles []FileInfo, severity Severity, endTimestamp int64) []FileInfo {
+	logFileMap := make(map[int64]FileInfo)
+	var timestamps int64sortable
+	for _, logFile := range logFiles {
+		if logFile.Details.Severity != severity {
+			continue
+		}
+
+		timestamp := logFile.Details.Time.UnixNano()
+		if timestamp <= endTimestamp {
+			logFileMap[timestamp] = logFile
+			timestamps = append(timestamps, timestamp)
+		}
+	}
+
+	// Sort the files in reverse order so we will fetch the newest first.
+	sort.Sort(sort.Reverse(timestamps))
+
+	result := []FileInfo{}
+	for _, timestamp := range timestamps {
+		result = append(result, logFileMap[timestamp])
+	}
+	return result
+}
+
+// FetchEntriesFromFiles fetches all available logs on disk that are of the
+// level of severity (or worse) and are between the start and end times. It will
+// stop reading in new files if the EntiresCutoff is exceeded and no more files
+// will be retrieved. The logs entries returned will be in decreasing order,
+// with the closest log to the start time being the first entry.
+func FetchEntriesFromFiles(severity Severity, startTimestamp, endTimestamp int64) ([]proto.LogEntry, error) {
+	logFiles, err := ListLogFiles()
+	if err != nil {
+		return nil, err
+	}
+
+	selectedFiles := selectFiles(logFiles, severity, endTimestamp)
+
+	entries := []proto.LogEntry{}
+	for _, file := range selectedFiles {
+		newEntries, entryBeforeStart, err := readAllEntriesFromFile(file, startTimestamp, endTimestamp)
+		if err != nil {
+			return nil, err
+		}
+		entries = append(entries, newEntries...)
+		if len(entries) >= EntiresCutoff {
+			break
+		}
+		if entryBeforeStart {
+			// Stop processing files that won't have any timestamps after
+			// startTime.
+			break
+		}
+	}
+	return entries, nil
+}
+
+// readAllEntriesFromFile reads in all log entries from a given file that are
+// between the start and end times and returns the entries in the reverse order,
+// from newest to oldest. It also returns a flag that denotes if any timestamp
+// occurred before the startTime to ensure inform the caller that no more log
+// files need to be processed.
+func readAllEntriesFromFile(file FileInfo, startTimestamp, endTimestamp int64) ([]proto.LogEntry, bool, error) {
+	reader, err := GetLogReader(file.Name, false)
+	defer reader.Close()
+	if reader == nil || err != nil {
+		return nil, false, err
+	}
+	entries := []proto.LogEntry{}
+	decoder := NewEntryDecoder(reader)
+	entryBeforeStart := false
+	for {
```

Comment:
I'm not sure why it was so high. Maybe there were issues at google with large numbers of files in one directory (google's initial log aggregation system was to export the log directory from each machine over nfs). Lowering the limit to something that can be held in memory seems like a good idea to me.